python/metashape_workflow_functions.py

```diff
@@ -4,1025 +4,1139 @@
-    return None
-
-
-#### Functions for each major step in Metashape
-
-
-def project_setup(cfg, config_file):
-    """
-    Create output and project paths, if they don't exist
-    Define a project ID based on specified project name and timestamp
-    Define a project filename and a log filename
-    Create the project
-    Start a log file
-    """
-
-    # Make project directories (necessary even if loading an existing project because this workflow saves a new project based on the old one, leaving the old one intact
-    if not os.path.exists(cfg["output_path"]):
-        os.makedirs(cfg["output_path"])
-    if not os.path.exists(cfg["project_path"]):
-        os.makedirs(cfg["project_path"])
-
-    ### Set a filename template for project files and output files based on the 'run_name' key of the config YML
-    ## BUT if the value for run_name is "from_config_filename", then use the config filename for the run name.
-
-    run_name = cfg["run_name"]
-
-    if run_name == "from_config_filename" or run_name ==  "":
-        file_basename = os.path.basename(config_file)  # extracts file base name from path
-        run_name, _ = os.path.splitext(file_basename)  # removes extension
+class MetashapeWorkflow:
+
+    def __init__(self, config_file):
+        self.config_file = config_file
+        self.doc = None
+        self.log_file = None
+        self.run_id = None
+        self.cfg = None
+        self.read_yaml()
+        self.sep = "; "
+
+    def convert_objects(self, a_dict):
+        """
+        Convert strings that refer to metashape objects (e.g. "Metashape.MoasicBlending") into metashape objects
+
+        Based on
+        https://stackoverflow.com/a/25896596/237354
+        """
+        for k, v in a_dict.items():
+            if not isinstance(v, dict):
+                if isinstance(v, str):
+                    # TODO look for Metashape.
+                    if (
+                        v
+                        and "Metashape" in v
+                        and not ("path" in k)
+                        and not ("project" in k)
+                        and not ("name" in k)
+                    ):  # allow "path" and "project" and "name" keys (e.g. "photoset_path" and "run_name") from YAML to include "Metashape" (e.g., Metashape in the filename)
+                        a_dict[k] = eval(v)
+                elif isinstance(v, list):
+                    # skip if no item in list have metashape, else convert string to metashape object
+                    if any("Metashape" in item for item in v):
+                        a_dict[k] = [eval(item) for item in v if ("Metashape" in item)]
+            else:
+                self.convert_objects(v)
+
+    def read_yaml(self):
+        with open(self.config_file, "r") as ymlfile:
+            self.cfg = yaml.load(ymlfile, Loader=yaml.SafeLoader)
+
+        # TODO: wrap in a Try to catch errors
+        self.convert_objects(self.cfg)
+
+    def stamp_time(self):
```

Comment:
I generally prefer to change methods that don't use `self` to functions, defined at the top of the file or in another one. To me, this makes it clear that there aren't any modifications or dependencies on the state of `self`. You could also use the `@staticmethod` decorator to convey this explicitly.